File: lib/test/t_LeastSquaresExpansion_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of the LeastSquaresExpansion class
 *
 *  Copyright 2005-2023 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    const Point expectedCoefficientsLinear = {3.5, 1.62542, 0, 0, 0, 0, 0, -0.594723, 0, 0, -1.29064, 0, 0, 0, 0, 1.37242, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1.95229, 0, 0, 0, 0, 0.194929, 0, 0, 0, 0, -1.08975, 0, 0, 0, 0, 0, 0, 0, 0, 0.409177, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1.35741, 0, 0, 0, 0, 0, 0, -0.0126684, 0, 0, 0, 0, 0.164588, 0, 0, 0, 0, 0, 0, 0, 0, -0.324901, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -0.33939, 0, 0, 0, 0, 0, 0, 0, 0, 0.00046142, 0, 0, 0, 0, -0.0106965, 0, 0, 0, 0, 0, 0, 0, 0, 0.0490707, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.0459147, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    const Point expectedCoefficientsCondensed = {3.5, 1.62542, -0.594723, -1.29064, 1.37242, -1.95229, 0.194929, -1.08975, 0.409177, 1.35741, -0.0126684, 0.164588, -0.324901, -0.33939, 0.00046142, -0.0106965, 0.0490707, 0.0459147};
    const Indices condensedIndices = {0, 1, 7, 10, 15, 30, 35, 40, 49, 77, 84, 89, 98, 156, 165, 170, 179, 275};
    const Point expectedCoefficientsHyper = {3.5, 1.62542, 0, 0, 0, -0.594723, 0, -1.29064, 0, 0, 0, 0, 0, 0, -1.95229, 0, 0.194929, 0, 0, 0, 0, 0, 0, 1.37242, 0, 0, 1.35741, 0, -0.0126684, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -0.33939, 0, 0, 0, 0, 0, 0, 0.409177, 0, 0.00046142, 0, 0, 0, -1.08975, 0, 0, 0, 0, 0, 0.0459147, 0, 0, 0, 0, 0, 0, 0};
    // Problem parameters
    UnsignedInteger dimension = 3;
    Scalar a = 7.0;
    Scalar b = 0.1;
    // Create the Ishigami function
    Description inputVariables(dimension);
    inputVariables[0] = "xi1";
    inputVariables[1] = "xi2";
    inputVariables[2] = "xi3";
    Description formula(1);
    formula[0] = (OSS() << "sin(xi1) + (" << a << ") * (sin(xi2)) ^ 2 + (" << b << ") * xi3^4 * sin(xi1)");
    SymbolicFunction model(inputVariables, formula);

    // Create the input distribution
    Collection<Distribution> marginals(dimension);
    marginals[0] = Uniform(-M_PI, M_PI);
    marginals[1] = Uniform(-M_PI, M_PI);
    marginals[2] = Uniform(-M_PI, M_PI);
    ComposedDistribution distribution(marginals);

    // Create the orthogonal basis
    Collection<OrthogonalUniVariatePolynomialFamily> polynomialCollection(dimension);
    polynomialCollection[0] = LegendreFactory();
    polynomialCollection[1] = LegendreFactory();
    polynomialCollection[2] = LegendreFactory();

    LinearEnumerateFunction enumerateFunction(dimension);
    OrthogonalProductPolynomialFactory productBasis(polynomialCollection, enumerateFunction);

    // Create the doe
    UnsignedInteger degree = 10;
    UnsignedInteger basisSize = enumerateFunction.getBasisSizeFromTotalDegree(degree);
    UnsignedInteger samplingSize = 8192;
    UnsignedInteger marginalSize = degree + 5;

    Collection<WeightedExperiment> doeList;
    RandomGenerator::SetSeed(0);
    doeList.add(LowDiscrepancyExperiment(LowDiscrepancySequence(SobolSequence()), distribution, samplingSize));
    doeList.add(GaussProductExperiment(distribution, Indices(distribution.getDimension(), marginalSize)));
    for (UnsignedInteger i = 0; i < doeList.getSize(); ++i)
    {
      // Sampling
      Point weights;
      Sample inputSample(doeList[i].generateWithWeights(weights));
      const Scalar wMin = *std::min_element(weights.begin(), weights.end());
      const Scalar wMax = *std::max_element(weights.begin(), weights.end());
      Sample outputSample(model(inputSample));
      // Check the full constructors
      Description methodNames = {"SVD", "QR", "Cholesky"};
      for (UnsignedInteger n = 0; n < methodNames.getSize(); ++n)
      {
        {
          // Create the polynomial chaos algorithm using the full constructor
          LeastSquaresExpansion algo(inputSample, weights, outputSample, distribution, productBasis, basisSize, methodNames[n]);
          fullprint << "algo=" << algo << std::endl;
          fullprint << "algo=" << algo.__str__() << std::endl;
          algo.run();
          // Check the coefficients
          FunctionalChaosResult result(algo.getResult());
          Point coeffs(result.getCoefficients().asPoint());
          Point ref(coeffs.getSize());
          std::copy(expectedCoefficientsLinear.begin(), expectedCoefficientsLinear.begin() + coeffs.getSize(), ref.begin());
          Scalar err = (coeffs - ref).norm();
          const Scalar rtol = 5.0e-2;
          const Scalar atol = 5.0e-2;
          assert_almost_equal(err, 0.0, rtol, atol);
          // Check the function restriction
          algo.setActiveFunctions(condensedIndices);
          fullprint << "algo=" << algo.__str__() << std::endl;
          algo.run();
          {
            FunctionalChaosResult result(algo.getResult());
            Point coeffs(result.getCoefficients().asPoint());
            Point ref(coeffs.getSize());
            std::copy(expectedCoefficientsCondensed.begin(), expectedCoefficientsCondensed.begin() + coeffs.getSize(), ref.begin());
            Scalar err = (coeffs - ref).norm();
            const Scalar rtol = 5.0e-2;
            const Scalar atol = 5.0e-2;
            assert_almost_equal(err, 0.0, rtol, atol);
          } // reduced set of functions
        } // full constructor with weights
        if (wMin == wMax)
        {
          // Create the polynomial chaos algorithm using the full constructor
          LeastSquaresExpansion algo(inputSample, outputSample, distribution, productBasis, basisSize, methodNames[n]);
          fullprint << "algo=" << algo << std::endl;
          fullprint << "algo=" << algo.__str__() << std::endl;
          algo.run();
          // Check the coefficients
          FunctionalChaosResult result(algo.getResult());
          Point coeffs(result.getCoefficients().asPoint());
          Point ref(coeffs.getSize());
          std::copy(expectedCoefficientsLinear.begin(), expectedCoefficientsLinear.begin() + coeffs.getSize(), ref.begin());
          Scalar err = (coeffs - ref).norm();
          const Scalar rtol = 1.0e-3;
          const Scalar atol = 1.0e-3;
          assert_almost_equal(err, 0.0, rtol, atol);
        } // full constructor without weights
      } // Method
      {
        // Create the polynomial chaos algorithm using the simplified constructor
        LeastSquaresExpansion algo(inputSample, weights, outputSample, distribution);
        fullprint << "algo=" << algo << std::endl;
        fullprint << "algo=" << algo.__str__() << std::endl;
        algo.run();
        // Check the coefficients
        FunctionalChaosResult result(algo.getResult());
        Point coeffs(result.getCoefficients().asPoint());
        Point ref(coeffs.getSize());
        std::copy(expectedCoefficientsHyper.begin(), expectedCoefficientsHyper.begin() + coeffs.getSize(), ref.begin());
        Scalar err = (coeffs - ref).norm();
        const Scalar rtol = 1.0e-2;
        const Scalar atol = 1.0e-2;
        assert_almost_equal(err, 0.0, rtol, atol);
      } // Simplified with weights
      // Check the constructors assuming uniform weights
      if (wMin == wMax)
      {
        // Create the polynomial chaos algorithm using the simplified constructor
        LeastSquaresExpansion algo(inputSample, outputSample, distribution);
        fullprint << "algo=" << algo << std::endl;
        fullprint << "algo=" << algo.__str__() << std::endl;
        algo.run();
        // Check the coefficients
        FunctionalChaosResult result(algo.getResult());
        Point coeffs(result.getCoefficients().asPoint());
        Point ref(coeffs.getSize());
        std::copy(expectedCoefficientsHyper.begin(), expectedCoefficientsHyper.begin() + coeffs.getSize(), ref.begin());
        Scalar err = (coeffs - ref).norm();
        const Scalar rtol = 1.0e-2;
        const Scalar atol = 1.0e-2;
        assert_almost_equal(err, 0.0, rtol, atol);
      } // Simplified without weights
    } // doe
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}

File: python/test/t_LeastSquaresExpansion_std.py
#! /usr/bin/env python

import openturns as ot
from openturns.testing import assert_almost_equal
from openturns.usecases import ishigami_function

ot.TESTPREAMBLE()

im = ishigami_function.IshigamiModel()
expectedCoefficientsLinear = [
    3.5,
    1.62542,
    0,
    0,
    0,
    0,
    0,
    -0.594723,
    0,
    0,
    -1.29064,
    0,
    0,
    0,
    0,
    1.37242,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    -1.95229,
    0,
    0,
    0,
    0,
    0.194929,
    0,
    0,
    0,
    0,
    -1.08975,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0.409177,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1.35741,
    0,
    0,
    0,
    0,
    0,
    0,
    -0.0126684,
    0,
    0,
    0,
    0,
    0.164588,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    -0.324901,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    -0.33939,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0.00046142,
    0,
    0,
    0,
    0,
    -0.0106965,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0.0490707,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0.0459147,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
]
expectedCoefficientsCondensed = [
    3.5,
    1.62542,
    -0.594723,
    -1.29064,
    1.37242,
    -1.95229,
    0.194929,
    -1.08975,
    0.409177,
    1.35741,
    -0.0126684,
    0.164588,
    -0.324901,
    -0.33939,
    0.00046142,
    -0.0106965,
    0.0490707,
    0.0459147,
]
condensedIndices = [
    0,
    1,
    7,
    10,
    15,
    30,
    35,
    40,
    49,
    77,
    84,
    89,
    98,
    156,
    165,
    170,
    179,
    275,
]
expectedCoefficientsHyper = [
    3.5,
    1.62542,
    0,
    0,
    0,
    -0.594723,
    0,
    -1.29064,
    0,
    0,
    0,
    0,
    0,
    0,
    -1.95229,
    0,
    0.194929,
    0,
    0,
    0,
    0,
    0,
    0,
    1.37242,
    0,
    0,
    1.35741,
    0,
    -0.0126684,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    -0.33939,
    0,
    0,
    0,
    0,
    0,
    0,
    0.409177,
    0,
    0.00046142,
    0,
    0,
    0,
    -1.08975,
    0,
    0,
    0,
    0,
    0,
    0.0459147,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
]

# Extract the relevant elements of the Ishigami test-case
im = ishigami_function.IshigamiModel()
dimension = im.dim
model = im.model

marginals = [im.X1, im.X2, im.X3]
distribution = im.inputDistribution

# Create the orthogonal basis
polynomialCollection = [ot.LegendreFactory()] * dimension

enumerateFunction = ot.LinearEnumerateFunction(dimension)
productBasis = ot.OrthogonalProductPolynomialFactory(
    polynomialCollection, enumerateFunction
)

# Create the doe
degree = 10
basisSize = enumerateFunction.getBasisSizeFromTotalDegree(degree)
samplingSize = 8192
marginalSize = degree + 5

doeList = [
    ot.LowDiscrepancyExperiment(
        ot.LowDiscrepancySequence(ot.SobolSequence()), distribution, samplingSize
    ),
    ot.GaussProductExperiment(distribution, [marginalSize] * dimension),
]

for doe in doeList:
    # Sampling
    inputSample, weights = doe.generateWithWeights()
    wMin = min(weights)
    wMax = max(weights)
    outputSample = model(inputSample)
    # Check the full constructors
    for methodName in ["SVD", "QR", "Cholesky"]:
        # Create the polynomial chaos algorithm using the full constructor
        algo = ot.LeastSquaresExpansion(
            inputSample,
            weights,
            outputSample,
            distribution,
            productBasis,
            basisSize,
            methodName,
        )
        algo.run()
        # Check the coefficients
        result = algo.getResult()
        assert result.isLeastSquares()
        assert not result.involvesModelSelection()
        coeffs = result.getCoefficients().asPoint()
        ref = expectedCoefficientsLinear[: coeffs.getSize()]
        err = (coeffs - ref).norm()
        rtol = 1.0e-3
        atol = 1.0e-3
        assert_almost_equal(err, 0.0, rtol, atol)
        # Check the function restriction
        algo.setActiveFunctions(condensedIndices)
        print("algo=", algo)
        algo.run()
        result = algo.getResult()
        coeffs = result.getCoefficients().asPoint()
        ref = expectedCoefficientsCondensed[: coeffs.getSize()]
        err = (coeffs - ref).norm()
        rtol = 1.0e-3
        atol = 1.0e-3
        assert_almost_equal(err, 0.0, rtol, atol)
        if wMin == wMax:
            # Create the polynomial chaos algorithm using the full constructor
            algo = ot.LeastSquaresExpansion(
                inputSample,
                outputSample,
                distribution,
                productBasis,
                basisSize,
                methodName,
            )
            algo.run()
            # Check the coefficients
            result = algo.getResult()
            coeffs = result.getCoefficients().asPoint()
            ref = expectedCoefficientsLinear[: coeffs.getSize()]
            err = (coeffs - ref).norm()
            rtol = 1.0e-3
            atol = 1.0e-3
            assert_almost_equal(err, 0.0, rtol, atol)
    # Create the polynomial chaos algorithm using the simplified constructor
    algo = ot.LeastSquaresExpansion(inputSample, weights, outputSample, distribution)
    algo.run()
    # Check the coefficients
    result = algo.getResult()
    coeffs = result.getCoefficients().asPoint()
    ref = expectedCoefficientsHyper[: coeffs.getSize()]
    err = (coeffs - ref).norm()
    rtol = 1.0e-2
    atol = 1.0e-2
    assert_almost_equal(err, 0.0, rtol, atol)
    # Check the constructors assuming uniform weights
    if wMin == wMax:
        # Create the polynomial chaos algorithm using the simplified constructor
        algo = ot.LeastSquaresExpansion(inputSample, outputSample, distribution)
        algo.run()
        # Check the coefficients
        result = algo.getResult()
        coeffs = result.getCoefficients().asPoint()
        ref = expectedCoefficientsHyper[: coeffs.getSize()]
        err = (coeffs - ref).norm()
        rtol = 1.0e-2
        atol = 1.0e-2
        assert_almost_equal(err, 0.0, rtol, atol)
